Source/WTF/wtf/FunctionPtr.h

```diff
@@ -0,0 +1,228 @@
+    constexpr TypedFunctionPtr(Out(FASTCALL *ptr)(In...))
+        : m_ptr(encode(ptr))
+    { }
+#endif
+
+    Out operator()(In... in) const
+    {
+        ASSERT(m_ptr);
+        return (*get())(std::forward<In>(in)...);
+    }
+
+    constexpr Ptr get() const { return decode(m_ptr); }
+
+    template<PtrTag otherTag>
+    TypedFunctionPtr<otherTag, Out, In...> retagged() const
+    {
+        if (!m_ptr)
+            return TypedFunctionPtr<otherTag, Out, In...>();
+        return TypedFunctionPtr<otherTag, Out, In...>(decode(m_ptr));
+    }
+
+    constexpr void* executableAddress() const { return reinterpret_cast<void*>(m_ptr); }
+
+    template<PtrTag newTag>
+    void* retaggedExecutableAddress() const { return retagCodePtr<void*, tag, newTag>(m_ptr); }
+
+    void* untaggedExecutableAddress() const { return untagCodePtr<void*, tag>(m_ptr); }
+
+    explicit operator bool() const { return !!m_ptr; }
+    bool operator!() const { return !m_ptr; }
+
+    bool operator==(const TypedFunctionPtr& other) const { return m_ptr == other.m_ptr; }
+    bool operator!=(const TypedFunctionPtr& other) const { return m_ptr != other.m_ptr; }
+
+    TypedFunctionPtr& operator=(Ptr ptr)
+    {
+        m_ptr = encode(ptr);
+        return *this;
+    }
+
+    template<PtrTag otherTag>
+    TypedFunctionPtr& operator=(const TypedFunctionPtr<otherTag, Out, In...>& other)
+    {
+        m_ptr = encode(other.get());
+        return *this;
+    }
+
+    TypedFunctionPtr& operator=(std::nullptr_t)
+    {
+        m_ptr = nullptr;
+        return *this;
+    }
+
+protected:
+    enum AlreadyTaggedValueTag { AlreadyTaggedValue };
+    TypedFunctionPtr(AlreadyTaggedValueTag, void* ptr)
+        : m_ptr(bitwise_cast<Ptr>(ptr))
+    {
+        assertIsNullOrTaggedWith<tag>(ptr);
+    }
+
+    ALWAYS_INLINE static constexpr Ptr encode(Ptr ptr)
+    {
+        if constexpr (tag == CFunctionPtrTag)
+            return ptr;
+        if (!ptr)
+            return nullptr;
+        assertIsNullOrCFunctionPtr(ptr);
+        auto result = retagCodePtr<CFunctionPtrTag, tag>(ptr);
+        ASSERT_NULL_OR_VALID_CODE_POINTER(result);
+        return result;
+    }
+
+    ALWAYS_INLINE static constexpr Ptr decode(Ptr ptr)
+    {
+        if constexpr (tag == CFunctionPtrTag)
+            return ptr;
+        if (!ptr)
+            return nullptr;
+        ASSERT_NULL_OR_VALID_CODE_POINTER(ptr);
+        auto result = retagCodePtr<tag, CFunctionPtrTag>(ptr);
+        assertIsNullOrCFunctionPtr(result);
+        return result;
+    }
+
+    Ptr m_ptr;
+};
+
+static_assert(sizeof(TypedFunctionPtr<CFunctionPtrTag, int, int>) == sizeof(void*));
+#if COMPILER_SUPPORTS(BUILTIN_IS_TRIVIALLY_COPYABLE)
+static_assert(__is_trivially_copyable(TypedFunctionPtr<CFunctionPtrTag, int, int>));
+#endif
+
+template<PtrTag tag>
+class FunctionPtr : public TypedFunctionPtr<tag, void, int> {
+    using Base = TypedFunctionPtr<tag, void, int>;
+    using PtrType = void(*)(int);
```

Comment:
Why treat it as a function pointer at all, instead of just void*?

Reply:
Because the `TypedFunctionPtr` only handles function pointers.  Since `FunctionPtr` here is re-using `TypedFunctionPtr`'s implementation, it needs to play by those rules ... not that `FunctionPtr` really cares how the underlying bits are represented.